### PR TITLE
SOLID Approach implementation

### PR DESCRIPTION
**Pull request by sreena7h**

### CODE REVIEW

File: apis/client.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import requests
+
+from apis.interfaces import IAPIClient
+
+
+class NoAuthIAPIClient(IAPIClient):
+    def __init__(self, host):
+        self.host = host
+
+    def get_auth_header(self):
```
Comment (jibinkmathew94) [inline apis/client.py:11]: Doesnt this violate LISKOV Substitution?

File: apis/client.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        return None
+
+    def request(self, url: str, method: str,
+                data: dict = None, timeout: int = 5):
```
Comment (jibinkmathew94) [inline apis/client.py:15]: Should data be empty dict by default?

File: apis/interfaces.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from abc import ABC, abstractmethod
+
+
+class IAPIClient(ABC):
+
+    @abstractmethod
+    def request(self, url, method, data=None):
```
Comment (jibinkmathew94) [inline apis/interfaces.py:7]: Should you define expected data types here?

File: notion/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        pass
+
+
+class INotionAPIClient(ABC):
+    host = None
+    notion_page_id = None
+    api_client: IAPIClient = None
+
+    @abstractmethod
+    def get_page_content_from_response(self, response):
```
Comment (jibinkmathew94) [inline notion/interfaces.py:26]: Should you define expected data types here?

File: notion/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+from apis.interfaces import IAPIClient
+
+
+class INotionDataParser(ABC):
+    @abstractmethod
+    def parse(self, data: dict):
+        pass
+
+    @abstractmethod
+    def _format_text(self, text: str, value_type: str):
```
Comment (jibinkmathew94) [inline notion/interfaces.py:12]: Should private methods be defined in abstract classes, let's check on whether it is required

File: apis/client.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import json
+import requests
+
+from apis.interfaces import IAPIClient
+
+
+class NoAuthIAPIClient(IAPIClient):
+    def __init__(self, host):
+        self.host = host
```
Comment (jibinkmathew94) [inline apis/client.py:9]: Why this if you are not using this?

File: notion/helper.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        data = {
+            "page": {"id": self.notion_page_id},
+            "limit": 30,
+            "cursor": {"stack": []},
+            "chunkNumber": 0,
+            "verticalColumns": False
+        }
+        response = self.api_client.request(url, method='POST', data=data)
+        if not response:
+            response = self.api_client.request(url, method='POST', data=data)
```
Comment (jibinkmathew94) [inline notion/helper.py:65]: Why this?

File: notion/helper.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+class NotionAPIClient(INotionAPIClient):
+
+    host = os.environ.get('NOTION_HOST')
+    notion_page_id = os.environ.get('NOTION_PAGE_ID')
+
+    def __init__(self, api_client):
+        self.api_client = api_client
+
+    def get_page_content_from_response(self, response_data: dict):
```
Comment (jibinkmathew94) [inline notion/helper.py:50]: Should this be a private method?

File: notion/helper.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            text = self._format_text(text, value_type)
+            if text:
+                page.append(text)
+        page_string = '\n'.join(page)
+        return page_string
+
+
+class NotionAPIClient(INotionAPIClient):
+
+    host = os.environ.get('NOTION_HOST')
```
Comment (jibinkmathew94) [inline notion/helper.py:44]: Why is this fetched from the environment and not initiated? Can tests be written against this implementation?

File: notion/helper.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, api_client):
+        self.api_client = api_client
+
+    def get_page_content_from_response(self, response_data: dict):
+        data = response_data.get("recordMap", {}).get("block", {})
+        return data
+
+    def get_status_data(self):
+        url = os.path.join(self.host, 'api/v3/loadCachedPageChunk/')
+        data = {
```
Comment (jibinkmathew94) [inline notion/helper.py:56]: Would you need to construct this data for any other APIs in the future, if yes should this be a separate function?

File: notion/helper.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    def parse(self, data):
+        page = []
+        for _, data_value in data.items():
+            value = data_value.get('value')
+            value_type = value.get('type')
+            title = value.get('properties', {}).get('title', [['']])
+            text = self._parse_text(title)
+
+            if not text:
+                break
```
Comment (jibinkmathew94) [inline notion/helper.py:33]: Should this be break or continue, just asking, not going deep into logic

File: lamda_handler.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+def lambda_handler(event, context):
+    api_client = NoAuthIAPIClient(host=NOTION_HOST_URL)
+    notion = Notion(api_client=api_client)
+    status_data = notion.get_status_data()
+    parse_data = notion.parse(status_data)
+
+    sdk = WebClient(token=USER_TOKEN)
+    slack_client = SlackClient(sdk=sdk)
+    convo_history = slack_client.get_conversation_history(
```
Comment (jibinkmathew94) [inline lamda_handler.py:19]: Should this be not in the main function?

Comment (sreena7h) [reply lamda_handler.py:19]: sorry?

File: notion/helper.py (hunk 11 above)
Comment (sreena7h) [reply notion/helper.py:33]: Its break;

File: notion/helper.py (hunk 10 above)
Comment (sreena7h) [reply notion/helper.py:56]: Yes it should be

File: notion/helper.py (hunk 7 above)
Comment (sreena7h) [reply notion/helper.py:65]: Its due to an issue with Notion, First time they return null data, but its getting fixed in the very next call.

**Pull request closed without merging**